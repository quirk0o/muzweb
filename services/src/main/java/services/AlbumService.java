package services;

import model.Album;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.Assert;
import repositories.AlbumRepository;

import java.util.List;


@Service
public class AlbumService {

    @Autowired
    AlbumRepository albumRepository;

    public List<Album> getAllAlbums(int pageNumber, int pageSize) {
        List<Album> allAlbums = albumRepository.findAll();
        Assert.isTrue(pageNumber * pageSize <= allAlbums.size());
        return allAlbums.subList(pageNumber * pageSize, Math.min((pageNumber + 1) * pageSize, allAlbums.size()));
    }

    @Transactional
    public void saveAlbum(Album album) {
        albumRepository.save(album);
    }

    public Album getAlbumById(Long id) {
        Album album = albumRepository.findOne(id);
        Assert.notNull(album, "There is no such album");
        return album;
    }

    public Album voteForAlbum(Long id, int vote) {
        Album album = getAlbumById(id);
        album.vote(vote);
        saveAlbum(album);
        return album;
    }

    public List<Album> searchByname(String prefix) {
        return albumRepository.findByNameStartsWith(prefix);
    }

    public void removeAlbum(Album album) {
        albumRepository.delete(album);
    }
}