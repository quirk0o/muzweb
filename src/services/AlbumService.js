import API from './API'

const getAll = (pgNum, pgSize) => API.get(`/tracks/?pgNum=${pgNum}&pgSize=${pgSize}`)
const getOne = (id) => API.get(`/albums/${id}`)
const createAlbum = (album) => API.post(`/albums/add`, album)
const searchForAlbum = (prefix) => API.get(`/albums/search/${prefix}`)


export default {getAll, getOne, createAlbum, searchForAlbum}