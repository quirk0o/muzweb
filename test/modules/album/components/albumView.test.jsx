import React from 'react'
import {mount} from 'enzyme'
import {MemoryRouter} from 'react-router-dom'
import {Button, RatingIcon} from 'semantic-ui-react'

import AlbumService from '../../../../src/services/AlbumService'
import AlbumView from 'modules/album/components/albumView'

const mockAlbum = {
  id: 1,
  name: 'Album A',
  description: 'An album',
  releaseDate: '2017/06/07',
  author: {firstName: 'Bob', lastName: 'Marley'},
  tracks: [
    {id: 1, name: 'Track 1'},
    {id: 1, name: 'Track 2'},
    {id: 1, name: 'Track 3'},
    {id: 1, name: 'Track 4'}
  ]
}
let mockGetOne

jest.mock('semantic-ui-react', () => ({
  ...require.requireActual('semantic-ui-react'),
  Popup: ({children}) => children
}))

jest.mock('../../../../src/services/AlbumService', () => ({
  getOne: () => mockGetOne,
  voteForAlbum: jest.fn().mockImplementation(() => Promise.resolve())

}))

describe('<AlbumView />', function () {
  const wrapper = () => mount(
    <MemoryRouter>
      <AlbumView match={{params: {id: 1}}} />
    </MemoryRouter>
  )

  beforeEach(function () {
    mockGetOne = Promise.resolve(mockAlbum)
  })

  afterEach(function () {
    AlbumService.voteForAlbum.mockClear()
  })

  it('renders', function () {
    const subject = wrapper()
    return mockGetOne.then(() => {
      expect(subject.find('.albumView-main').render()).toMatchSnapshot()
    })
  })

  context('when vote is submitted', function () {
    it('registers vote', function () {
      const subject = wrapper()
      return mockGetOne.then(() => {
        subject
          .find(RatingIcon)
          .at(0)
          .find('.icon')
          .simulate('click')
        subject
          .find(Button)
          .filterWhere(button => button.text() === 'Vote')
          .find('button')
          .simulate('click')
        expect(AlbumService.voteForAlbum).toHaveBeenCalledWith(1, 1)
      })
    })
  })
})