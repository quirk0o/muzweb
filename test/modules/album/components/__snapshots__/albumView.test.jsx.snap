// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlbumView /> renders 1`] = `
<div
  class="ui center aligned container albumView-main"
>
  <h3
    class="ui header albumView-header"
  >
    Title: 
    Album A
    <div
      class="ui centered divided one column grid"
    >
      <div
        class="center aligned column"
      >
        <div
          class="albumView-ratingSegment"
        >
          <div>
            <div
              class="ui star rating albumView-rating"
              role="radiogroup"
            >
              <div
                aria-checked="false"
                aria-posinset="1"
                aria-setsize="10"
                class="icon"
                role="radio"
                tabindex="0"
              />
              <div
                aria-checked="false"
                aria-posinset="2"
                aria-setsize="10"
                class="icon"
                role="radio"
                tabindex="0"
              />
              <div
                aria-checked="false"
                aria-posinset="3"
                aria-setsize="10"
                class="icon"
                role="radio"
                tabindex="0"
              />
              <div
                aria-checked="false"
                aria-posinset="4"
                aria-setsize="10"
                class="icon"
                role="radio"
                tabindex="0"
              />
              <div
                aria-checked="false"
                aria-posinset="5"
                aria-setsize="10"
                class="icon"
                role="radio"
                tabindex="0"
              />
              <div
                aria-checked="false"
                aria-posinset="6"
                aria-setsize="10"
                class="icon"
                role="radio"
                tabindex="0"
              />
              <div
                aria-checked="false"
                aria-posinset="7"
                aria-setsize="10"
                class="icon"
                role="radio"
                tabindex="0"
              />
              <div
                aria-checked="false"
                aria-posinset="8"
                aria-setsize="10"
                class="icon"
                role="radio"
                tabindex="0"
              />
              <div
                aria-checked="false"
                aria-posinset="9"
                aria-setsize="10"
                class="icon"
                role="radio"
                tabindex="0"
              />
              <div
                aria-checked="false"
                aria-posinset="10"
                aria-setsize="10"
                class="icon"
                role="radio"
                tabindex="0"
              />
            </div>
          </div>
          <div>
            <button
              class="ui button albumView-button"
            >
              Vote
            </button>
          </div>
        </div>
      </div>
    </div>
  </h3>
  <h4
    class="ui header"
  >
    Author: 
    Bob Marley
  </h4>
  <h4
    class="ui header"
  >
    Release Date: 
    2017/06/07
  </h4>
  <div>
     
    An album
     
  </div>
  <div
    class="ui animated divided ordered selection list"
    role="list"
  >
    <a
      class="item"
      href="/tracks/1"
      role="listitem"
    >
      Track 1
    </a>
    <a
      class="item"
      href="/tracks/1"
      role="listitem"
    >
      Track 2
    </a>
    <a
      class="item"
      href="/tracks/1"
      role="listitem"
    >
      Track 3
    </a>
    <a
      class="item"
      href="/tracks/1"
      role="listitem"
    >
      Track 4
    </a>
  </div>
</div>
`;
